# Анализ базы данных стартапа

### Информация о базе данных
Необходимо проанализировать базу данных, в которых содержатся информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Задания

- Определить, количество книг, вышедших после 1 января 2000 года;
- Посчитать для каждой книги количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
- Выявить автора с самой высокой средней оценкой книг, при этом надо учитывать только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine 
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
#Выведем первые 5 строк каждой таблицы:
books1 = ''' SELECT *
              FROM books
          '''
books = pd.io.sql.read_sql(books1, con = engine)
display(books.head())

authors1 = ''' SELECT *
                FROM authors
          '''
authors = pd.io.sql.read_sql(authors1, con = engine)
display(authors.head())

publishers1 = ''' SELECT *
                   FROM publishers
          '''
publishers = pd.io.sql.read_sql(publishers1, con = engine)
display(publishers.head())

ratings1 = ''' SELECT *
                FROM ratings
          '''
ratings = pd.io.sql.read_sql(ratings1, con = engine)
display(ratings.head())

reviews1 = ''' SELECT *
                FROM reviews
          '''
reviews = pd.io.sql.read_sql(reviews1, con = engine)
display(reviews.head())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [3]:
# Выведем информацю о датасетах:
display(books.info())
display(authors.info())
display(publishers.info())
display(ratings.info())
display(reviews.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

**Задание 1:** *Посчитайте, сколько книг вышло после 1 января 2000 года;*

In [4]:
#Решение:
query_1 = ''' SELECT COUNT(book_id) as Количество_книг
                FROM books
               WHERE publication_date > '2000-01-01'
          '''
request_1 = pd.io.sql.read_sql(query_1, con = engine)
display(request_1)

Количество_книг
0              819

**Вывод:** Всего у нас 1000 книг в базе данных и из них 819 вышло после 2000го года, следовательно, большая часть книг можно назвать сравнительно новыми.

**Задание 2:** *Для каждой книги посчитайте количество обзоров и среднюю оценку;*

In [5]:
#Решение:
query_2 = ''' SELECT b.title AS Название_книги,
                     AVG(ra.rating) AS Средняя_оценка,
                     COUNT (DISTINCT(r.review_id)) AS Количество_обзоров
              FROM books AS b
              JOIN ratings AS ra ON ra.book_id = b.book_id
              LEFT JOIN reviews AS r ON r.book_id = b.book_id
              GROUP BY b.book_id
              ORDER BY COUNT (DISTINCT(r.review_id)) DESC;
          '''
request_2 = pd.io.sql.read_sql(query_2, con = engine)
display(request_2)

Название_книги  Средняя_оценка  \
0                              Twilight (Twilight  #1)        3.662500   
1                                  Water for Elephants        3.977273   
2                                     The Glass Castle        4.206897   
3    Harry Potter and the Prisoner of Azkaban (Harr...        4.414634   
4    The Curious Incident of the Dog in the Night-Time        4.081081   
..                                                 ...             ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel        3.666667   
996                            The Natural Way to Draw        3.000000   
997   The Cat in the Hat and Other Dr. Seuss Favorites        5.000000   
998                          Essential Tales and Poems        4.000000   
999  Disney's Beauty and the Beast (A Little Golden...        4.000000   

     Количество_обзоров  
0                     7  
1                     6  
2                     6  
3                     6  
4                     6  
..                  ...  
995                   0  
996                   0  
997                   0  
998                   0  
999                   0  

[1000 rows x 3 columns]

**Вывод:** Наибольшее количество обзоров у книги "Twilight (Twilight #1)"с рейтингом 3.66- 7.

**Задание 3:** *Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;*

In [6]:
#Решение:
query_3 = ''' SELECT MAX(p.publisher) as Издательство,
              COUNT (b.book_id) as Количество_книг
              FROM books as b
              LEFT JOIN publishers as p ON b.publisher_id=p.publisher_id
              WHERE b.num_pages > 50
              GROUP BY p.publisher
              ORDER BY COUNT(b.book_id) DESC
              LIMIT 1;
          '''
request_3 = pd.io.sql.read_sql(query_3, con = engine)
display(request_3)

Издательство  Количество_книг
0  Penguin Books               42

**Вывод:** Издательство Penguin Books выпустило 42 книги, с количеством страниц больше 50. 

**Задание 4:** *Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;*

In [7]:
#Решение:
query_4 = '''  SELECT author_1 AS Автор,
                     AVG (avg_rating) as Средний_рейтинг_книг
                FROM (SELECT b.book_id,
                             COUNT(DISTINCT r.rating_id) AS Количество_оценок,
                             AVG(r.rating) AS avg_rating,
                             a.author_id,
                             a.author AS author_1
                        FROM books as b
                   LEFT JOIN authors as a ON b.author_id=a.author_id
                   LEFT JOIN ratings as r ON r.book_id=b.book_id 
                    GROUP BY b.book_id, a.author_id  
                HAVING COUNT(r.rating_id) >= 50
                    ORDER BY AVG (r.rating) DESC)T
               GROUP BY Автор
            LIMIT 1;
          '''
request_4 = pd.io.sql.read_sql(query_4, con = engine)
display(request_4)

Автор  Средний_рейтинг_книг
0  J.K. Rowling/Mary GrandPré              4.283844

**Вывод:** J.K. Rowling/Mary GrandPré- автор с самой высокой средней оценкой книг (4.28). 

**Задание 5:** *Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.*

In [8]:
#Решение:
query_5 = ''' SELECT AVG(cnt) as Среднее_количество_обзоров
                FROM (SELECT COUNT(re.review_id) as cnt
                       FROM reviews as re
                      WHERE re.username IN (SELECT r.username
                                            FROM ratings as r
                                        GROUP BY r.username
                                          HAVING COUNT(r.rating_id) > 50)
                                          GROUP BY re.username) T;
          '''
request_5 = pd.io.sql.read_sql(query_5, con = engine)
display(request_5)

Среднее_количество_обзоров
0                   24.333333

**Вывод:** Среднее количество обзоров среди активных пользователей, поставивших больше 50 оценок, 24.33